### ISSUE TITLE
[Bug]: Wrong minor colorbar ticks in the presence of BoundaryNorm

### ISSUE THREAD

**Issue by zxdawn** [matplotlib/matplotlib#22233]
### Bug summary

The minor colorbar ticks are messed up for matplotlib3.5+. But, it works well for v3.4.3.

### Code for reproduction

```python
import matplotlib
import numpy as np
import matplotlib.pyplot as plt

bounds = np.array([0, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 0.97, 0.98, 0.99, 1.0])
norm = matplotlib.colors.BoundaryNorm(boundaries=bounds, ncolors=256)

x = y = np.array([-10, -5, 0, 5, 10])
np.random.seed(20)
data = np.random.uniform(0,1,(4,4))

fig = plt.figure()
ax = fig.add_subplot()
m = ax.pcolormesh(x, y, data, norm=norm)
cb = fig.colorbar(m)
cb.minorticks_on()
```

### Actual outcome

![image](https://user-images.githubusercontent.com/30388627/149580197-cd3ad92d-c6e9-4155-8fd2-f798c1b56f43.png)

### Expected outcome

![image](https://user-images.githubusercontent.com/30388627/149579946-f23e5273-0bcb-43a2-a3e2-0c7bb5691edd.png)

### Operating system

Ubuntu

### Matplotlib Version

3.5.1

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

3.9.7

### Jupyter version

3.2.4

### Installation

conda

**Comment by jklymak** [comment]
Colorbars had a major change in 3.5, so sorry this unexpectedly changed for you 

Otoh I don't see how the minor ticks were particularly better in the old version.  What is the goal of having minor ticks between boundaries?  I'm not sure what the goal is with the minor ticks with a boundary norm.

**Comment by lukelbd** [comment]
I suppose you're right that `AutoMinorLocator` is not particularly useful in combination with `BoundaryNorm`. Minor ticks are really only useful here to indicate level boundaries that don't have major ticks/labels.

This example that adds minor ticks to level boundaries with `cb.minorlocator = FixedLocator(boundaries)` seems to work fine (since it has no underscore, I assume `cb.minorlocator` is part of the public API, but correct me if I'm wrong):

```python
import numpy as np
import matplotlib.colors as mcolors
import matplotlib.ticker as mticker
import matplotlib.pyplot as plt

crf_bounds = np.array([0, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 0.97, 0.98, 0.99, 1.0])
norm = mcolors.BoundaryNorm(boundaries=crf_bounds, ncolors=256)

x = y = np.array([-10, -5, 0, 5, 10])
np.random.seed(20)
data = np.random.uniform(0,1,(4,4))

fig = plt.figure()
ax = fig.add_subplot()
m = ax.pcolormesh(x, y, data, cmap='RdYlGn', norm=norm)
cb = fig.colorbar(m)
cb.minorlocator = mticker.FixedLocator(crf_bounds)
cb.update_ticks()
```
![tmp](https://user-images.githubusercontent.com/19657652/149604566-13b9b4f7-8215-446c-864e-028befdf42bb.png)


However, I'd argue this issue is still worth fixing, just because that default behavior (while seldom triggered) is quite bizarre, and I can easily imagine a scenario where someone enables minor colorbar ticks by default on all their plots and gets this result when they happen to use a `BoundaryNorm`.

Further, this issue seems to crop up when other locators are used for minor ticks -- not just `AutoMinorLocator`. Here's an example using `mticker.MultipleLocator` for the minor ticks:

```python
import numpy as np
import matplotlib.colors as mcolors
import matplotlib.ticker as mticker
import matplotlib.pyplot as plt

crf_bounds = np.array([0, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 0.97, 0.98, 0.99, 1.0])
norm = mcolors.BoundaryNorm(boundaries=crf_bounds, ncolors=256)

x = y = np.array([-10, -5, 0, 5, 10])
np.random.seed(20)
data = np.random.uniform(0,1,(4,4))

fig = plt.figure()
ax = fig.add_subplot()
m = ax.pcolormesh(x, y, data, cmap='RdYlGn', norm=norm)
cb = fig.colorbar(m)
fig.patch.set_alpha(1)
cb.minorlocator = mticker.FixedLocator(crf_bounds)
cb.minorlocator = mticker.MultipleLocator(0.2)
cb.update_ticks()
```

![tmp](https://user-images.githubusercontent.com/19657652/149604805-8b86c82c-5fbe-4500-b3a2-7f4d8c43980c.png)

**Comment by lukelbd** [comment]
Yeah I guess you're right that neither the old or new behaviors are particularly great. Probably the best use case for minor ticks in this situation would be to place minor ticks on level boundaries that don't have major ticks. Maybe some clause could be added to `minorticks_on` that simply uses the norm boundaries for minor ticks when the normalizer is `BoundaryNorm`.

This below example works, though not sure if `cb.minorlocator` is meant to be part of the public API, and I see that the `update_ticks` docstring warns users that they shouldn't have to use it.

```python
import matplotlib.colors as mcolors
import matplotlib.ticker as mticker
import numpy as np
import matplotlib.pyplot as plt

crf_bounds = np.array([0, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 0.97, 0.98, 0.99, 1.0])
norm = mcolors.BoundaryNorm(boundaries=crf_bounds, ncolors=256)

x = y = np.array([-10, -5, 0, 5, 10])
np.random.seed(20)
data = np.random.uniform(0,1,(4,4))

fig = plt.figure()
ax = fig.add_subplot()
m = ax.pcolormesh(x, y, data, cmap='RdYlGn', norm=norm)
cb = fig.colorbar(m)
fig.patch.set_alpha(1)
cb.minorlocator = mticker.FixedLocator(crf_bounds)
cb.update_ticks()
```

![tmp](https://user-images.githubusercontent.com/19657652/149605527-7bbc62b3-9f34-4794-a7fd-1656ed96b7f6.png)

**Comment by jklymak** [comment]
I think changing the minor locator to catch every boundary missed by the major locator is a reasonable default, at least if the colorbar is set to "uniform".  There is already special casing for boundary norm, so adding something to give a reasonable minor tick seems doable, but i would have to look again.

**Comment by jklymak** [comment]
Looking at the code, the locator and formatter selection on colorbars is still a bit chaotic, with some setting done on the underlying axes and some seeming to be done on the colorbar itself.  

Fixing this is relatively easy, but requires some API work.  I think we should have either 1) `cb.set_major_locator` and `cb.set_minor_locator`, or 2) promote `cb.locator` and c`cb.minor_locator` to properties with setter and getter.  If the user sets these manually, they need to call `update_ticks` for them to be properly passed down to the axes/axis objects.  

I guess I slightly favour 1) as more compatible with normal axes objects.  i.e. cb.set_minor_locator would just pass down to `cb.ax._long_axis.set_minor_locator`.   I guess we could do 1 _and_ 2, so long as we made it easy to change the code in just one underlying place...
